varname.py

```diff
@@ -461,16 +484,88 @@ def __repr__(self) -> str:
         return (f"<{self.__class__.__name__} "
                 f"(name={self.name!r}, value={self.value!r})>")
 
-def _get_frame(caller: int) -> FrameType:
-    """Get the frame at `caller` depth"""
+def _debug(msg: str, frame: Optional[FrameType] = None) -> None:
+    """Print the debug message"""
+    if not DEBUG:
+        return
+    if frame is not None:
+        frameinfo = inspect.getframeinfo(frame)
+        msg = (f'{msg} [In {frameinfo.function!r} at '
+               f'{frameinfo.filename}:{frameinfo.lineno}]')
+    sys.stderr.write(f'[{__name__}] DEBUG: {msg}\n')
+
+def _get_executing(
+        caller: int,
+        ignore: Optional[
+            List[Union[ModuleType, Tuple[ModuleType, str]]]
+        ] = None
+) -> FrameType:
+    """This function makes sure that we get the desired frame,
+    by caller and ignroe.
+
+    The caller is the desired stack we want after excluding the frames we want
+    to ignore. By default, this module (varname) and the builtin modules will
+    be ignored.
+
+    The frame_index defaults to 1 to exclude the calls from inside this module.
+    """
+    # Use loop instead of recursion to avoid creating additional stacks
     try:
-        return sys._getframe(caller + 1)
+        # We are at least skipping 2 frames:
+        # one is this function, the other is any varname API that uses this
+        # Don't bother to test them, just simply skip.
+        frame_index = 2
+
+        while caller > 0:
+            frame = sys._getframe(frame_index)
+            # loot at next frame anyway at next iteration
+            frame_index += 1
+            module = inspect.getmodule(frame)
+            exect = executing.Source.executing(frame)
```

Comment:
This is somewhat expensive, try to call it only when you need to get the node, i.e. when you return at the end. For just the qualname use `executing.Source.for_frame(frame).code_qualname(frame.f_code)`. Even that should only be when you actually need the qualname.